design-proposals/klip-47-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,122 @@
+
+For ksqlDB, the setup is quite differently. Specifically, user could create the database based on 
+the topics stored in Kafka as the source of truth. The defined database could be mutable by 
+using `INSERT` command to populate data into it, which will then pollute the original data source.
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor 
+ready for interactive queries to access. To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will materialize the input source topic into a stream 
+instance which would generate a RocksDB instance for interactive queries.
+
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
+* `DESCRIBE ... EXTENDED` should show source streams and tables as read-only  
+* Add materialization of the `source table` to make it pull-query accessible
+* Make source stream/table immutable from KSQL by rejecting insertion requests
+ 
+
+## What is not in scope
+* We will not add role-based write permission for source stream/table in v1, even though it seems to be useful.
+
+## Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
+* createStream
+* createStreamAs
+* createTable
+* createTableAs
+
+Then inside `AstBuilder#visitCreateStream` and `AstBuilder#visitCreateTable`, the built classes for stream/table 
+instances would be `CreateStream` and `CreateTable`, which are the only two subclasses of `CreateSource`. It makes sense 
+to add the `SOURCE` as a flag into `CreateSource`, and let `AstBuilder` pass the flag into the struct 
+for later stage reference of read-only.
+
+In addition, when user calls `DESCRIBE ... EXTENDED`, the read-only attribute shall be displayed for stream/table. 
+
+### Make Source Stream/Table Read-only
+We plan to add restriction check into the `DistributingExecutor#throwIfInsertOnReadOnlyTopic` or create a similar function, 
+which could verify whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+### Source Table Materialization
+In the call `EngineExecutor#execute`, we will check whether given plan has a `queryPlan` before deciding to proceed 
+creating a persistent query. To bypass this check, we need to provide a concrete query plan within the input plan. We would 
+inject a dummy query plan in `EngineExecutor#plan` phase, when the statement is a CreateTable statement with `SOURCE` tag.
+
+In addition, when user tries to show all the running queries, the materialization query should also be displayed as it takes 
+part of the computation resource. This would be a special type of running query without `SELECT` clause.
```

Comment:
This is an interesting aspect. I agree to the sentiment, but we should not display it as a "query" to the user IMHO. That might be confusing. It should be clear for the user that it's a running "source table program" (for the lack of a better term).

Reply:
Piggybacking off Matthias's comment, will we allow users to interact with this query? For example, will it expose a query ID like other queries? How will we prevent users from terminating the query if so?

Reply:
I don't think that there would be query ID -- the underlying program should be bound to the table IMHO, so user can `DROP` the table (and this will terminate the program, too).

Reply:
That makes sense.